projects/com.oracle.truffle.llvm/src/com/oracle/truffle/llvm/LLVM.java

```diff
@@ -75,9 +80,44 @@
             public CallTarget parse(Source code, Node context, String... argumentNames) {
                 Node findContext = LLVMLanguage.INSTANCE.createFindContextNode0();
                 LLVMContext llvmContext = LLVMLanguage.INSTANCE.findContext0(findContext);
-                ParserResult parserResult = parseFile(code.getPath(), llvmContext);
-                llvmContext.getFunctionRegistry().register(parserResult.getParsedFunctions());
-                return parserResult.getMainFunction();
+
+                if (code.getMimeType() == LLVMLanguage.LLVM_MIME_TYPE) {
+                    ParserResult parserResult = parseFile(code.getPath(), llvmContext);
+                    llvmContext.getFunctionRegistry().register(parserResult.getParsedFunctions());
+                    return parserResult.getMainFunction();
+                } else if (code.getMimeType() == LLVMLanguage.SULONG_LIBRARY_MIME_TYPE) {
+                    final List<CallTarget> mainFunctions = new ArrayList<>();
+
+                    final SulongLibrary library = new SulongLibrary(new File(code.getPath()));
+
+                    try {
+                        library.readContents(dependentLibrary -> {
+                            throw new UnsupportedOperationException();
+                        }, source -> {
+                            ParserResult parserResult;
+                            try {
+                                parserResult = parseString(source.getCode(), llvmContext);
+                            } catch (IOException e) {
+                                throw new UncheckedIOException(e);
+                            }
+
+                            llvmContext.getFunctionRegistry().register(parserResult.getParsedFunctions());
+                            if (parserResult.getMainFunction() != null) {
```

Comment:
Can this be `null` in the current implementation? In the visitor you create `Truffle.getRuntime().createCallTarget(RootNode.createConstantNode(null))` if there is no main function.